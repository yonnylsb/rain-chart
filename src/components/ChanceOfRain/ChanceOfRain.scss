@import '../../common/styles/colors.scss';

.cor {
    &__y-axis {
        transform: matrix(1,0,0,1,30,-50);
        font-size: 10px;
        &__text{
            transform: rotate(-90deg) translate(-290px, -48px);
            font-size: 12px;
            fill: $blue-1;
        }
        &__line{
            stroke:$blue-1;
            stroke-width:1;
        }
    }
    
    &__x-axis {
        font-size: 10px;
        transform: matrix(1,0,0,1,0,-30);
        &__text{
            font-size: 12px;
            fill: $blue-1;
        }
        &__line{
            stroke:$blue-1;
            stroke-width:1;
        }
    }
    &__upper {
        fill: transparent;
        stroke: $blue-3;
        transform: matrix(1,0,0,1,0,-50);
    }
    &__med {
        fill: transparent;
        stroke: $red-1;
        transform: matrix(1,0,0,1,0,-50);
    }
    &__lower {
        fill: transparent;
        stroke: $green-1;
        transform: matrix(1,0,0,1,0,-50);
    }
}