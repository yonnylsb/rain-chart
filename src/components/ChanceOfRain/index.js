import React from 'react';
import chanceOfRain from '../../common/constants/helpers';
import './ChanceOfRain.scss';

function ChanceOfRain(props){
    let svgWidth = 800;
    let svgHeight = 460;
    let yAxisX = 0;
    let xAxisY = 460;
    let renderYaxis = [];
    let renderXaxis = [];

    let daysValues = props.daysValues;
    let pressureGraph = props.pressureGraph;
    let temperatureGraph= props.temperatureGraph;

    let minX = props.minX;
    let maxX = props.maxX;
    let stepX = props.stepX;
    let textX = props.textX;

    let minY = props.minY;
    let maxY = props.maxY;
    let stepY = props.stepY;
    let textY = props.textY;

    let upper = [];
    let medMpah = [];
    let lower = [];
    
    let calculateStepY = (svgHeight-60) / ((maxY-minY)/stepY);
    let calculateStepX = (svgWidth-60) / ((maxX-minX)/stepX);

    if(Number.isInteger(minY) && maxY && stepY){
        let yAxisY = svgHeight;
        for(let i = props.minY; i <= props.maxY; i=i+props.stepY){
            renderYaxis.push(<text y={yAxisY} x={yAxisX} key={`axisX${yAxisY}`}>{i}</text>);
            yAxisY = yAxisY - calculateStepY;
        }
    }

    if(Number.isInteger(minX) && maxX && stepX){
        let xAxisX = 0;
        let idx = 0;
        for(let i = props.minX; i <= (props.maxX); i=i+props.stepX){
            renderXaxis.push(<text y={xAxisY} x={xAxisX} key={`axisX${xAxisX}`}>{i?i:""}</text>);

            if(xAxisX > 0){
                let parseChanceOfRain = chanceOfRain(pressureGraph, temperatureGraph, daysValues[idx-1]);
                upper.push([xAxisX, (svgHeight - 4*parseChanceOfRain[0])]);
                medMpah.push([xAxisX, (svgHeight - 4*parseChanceOfRain[1])]);
                lower.push([xAxisX, (svgHeight - 4*parseChanceOfRain[2])]);
            }
            xAxisX = xAxisX + calculateStepX;
            idx++;
        }
    }

    function parseNodes(arr){
        return arr.map((item, index)=>{
            return index ? `L ${item[0]} ${item[1]}`: `M ${item[0]} ${item[1]}`;
        }).join(" ");
    }

    return (
        <svg 
            viewBox={`0 0 ${svgWidth} ${svgHeight}`}
        >
            <path
                className="cor__upper"
                d={parseNodes(upper)}
            />
            <path
                className="cor__med"
                d={parseNodes(medMpah)}
            />
            <path
                className="cor__lower"
                d={parseNodes(lower)}
            />
            <line 
                x1="50" y1="0" x2="50" y2="410"
                className="cor__y-axis__line"
            />
            <line 
                x1="50" y1="410" x2="780" y2="410" 
                className="cor__x-axis__line"
            />
            <text
                x={20}
                y={60}
                className="cor__y-axis__text"
            >
                {textY}
            </text>
            <text
                x={400}
                y={450}
                className="cor__x-axis__text"
            >
                {textX}
            </text>
            <g className="cor__y-axis">
                {renderYaxis}
            </g>
            <g className="cor__x-axis">
                {renderXaxis}
            </g>
        </svg>
    );
}


export default ChanceOfRain;