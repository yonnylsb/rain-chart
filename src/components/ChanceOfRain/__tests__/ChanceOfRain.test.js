import React from 'react';
import renderer from 'react-test-renderer';
import ChanceOfRain from '../index';

test('ChanceOfRain should render', () => {
    const component = renderer.create(
        <ChanceOfRain></ChanceOfRain>
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
