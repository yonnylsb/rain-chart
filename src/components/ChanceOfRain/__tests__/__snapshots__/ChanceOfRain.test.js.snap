// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChanceOfRain should render 1`] = `
<svg
  viewBox="0 0 800 460"
>
  <path
    className="cor__upper"
    d=""
  />
  <path
    className="cor__med"
    d=""
  />
  <path
    className="cor__lower"
    d=""
  />
  <line
    className="cor__y-axis__line"
    x1="50"
    x2="50"
    y1="0"
    y2="410"
  />
  <line
    className="cor__x-axis__line"
    x1="50"
    x2="780"
    y1="410"
    y2="410"
  />
  <text
    className="cor__y-axis__text"
    x={20}
    y={60}
  />
  <text
    className="cor__x-axis__text"
    x={400}
    y={450}
  />
  <g
    className="cor__y-axis"
  />
  <g
    className="cor__x-axis"
  />
</svg>
`;
